### PR TITLE
Store a new Region value every time we create a new region variable

### PR DESCRIPTION
**Pull request by Nashenas88**
Paired with @spastorino to walk through this and implement #44870.

### CODE REVIEW

File: src/librustc_mir/transform/nll.rs [hunk 1]
```diff
@@ -143,4 +147,9 @@ impl MirPass for NLL {
             let _results = visitor.into_results();
         })
     }
-}
\ No newline at end of file
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
```
Comment (Nashenas88) [inline src/librustc_mir/transform/nll.rs]: We couldn't keep the derive for `Hash` because of `FxHashSet`. It errors out saying that `HashSet` itself doesn't implement `Hash`.

File: src/librustc_mir/transform/nll.rs [hunk 2]
```diff
@@ -29,15 +31,17 @@ impl<'a, 'gcx, 'tcx> NLLVisitor<'a, 'gcx, 'tcx> {
         NLLVisitor {
             infcx,
             lookup_map: HashMap::new(),
+            regions: vec![],
         }
     }
 
     pub fn into_results(self) -> HashMap<RegionVid, Lookup> {
```
Comment (Nashenas88) [inline src/librustc_mir/transform/nll.rs]: Should we output this as a tuple, i.e. `(HashMap<RegionVid, Lookup>, Vec<Region>)`? Or do we only worry about that when we need to start using the output?

**Comment by Nashenas88** [comment]
Should we actually fork off of #44845 or will it be possible to merge this in without trouble?

File: src/librustc_mir/transform/nll.rs [hunk 3]
```diff
@@ -15,12 +15,14 @@ use rustc::mir::{Mir, Location, Rvalue, BasicBlock, Statement, StatementKind};
...
 use rustc::mir::transform::{MirPass, MirSource};
 use rustc::infer::{self, InferCtxt};
+use rustc::util::nodemap::FxHashSet;
 use syntax_pos::DUMMY_SP;
 use std::collections::HashMap;
 
 #[allow(dead_code)]
 struct NLLVisitor<'a, 'gcx: 'a + 'tcx, 'tcx: 'a> {
     lookup_map: HashMap<RegionVid, Lookup>,
+    regions: Vec<Region>,
```
Comment (arielb1) [inline src/librustc_mir/transform/nll.rs]: Can you use a `rustc_data_structures::indexed_vec::IndexVec` here? usize indexes suck.

File: src/librustc_mir/transform/nll.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_mir/transform/nll.rs]: Hopefully this will not have to be `Hash`.

File: src/librustc_mir/transform/nll.rs [hunk 4]
```diff
@@ -143,4 +148,25 @@ impl MirPass for NLL {
...
\ No newline at end of file
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+struct Region {
+    points: FxHashSet<Location>,
+}
+
+#[derive(Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd, Debug)]
+pub struct RegionIndex(pub u32);
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/nll.rs]: Hmm, it'd be nice if we could use [the `newtype_idx!` macro](https://github.com/rust-lang/rust/blob/0e6f4cf51cd3b799fb057956f8e733d16605d09b/src/librustc/mir/mod.rs#L46-L68) for this. How about moving that macro into `librustc_data_structures`, so that everybody can use it?

To do that, you would move it into the [`indexed_vec`](https://github.com/rust-lang/rust/blob/0e6f4cf51cd3b799fb057956f8e733d16605d09b/src/librustc_data_structures/indexed_vec.rs) module and put a `#[macro_export]` annotation on it, so that it gets exported across crates (I think that's the right annotation, anyway). Then we have to tag `extern crate rustc_data_structures` in `src/librustc_mir/lib.rs` with `#[macro_use]` also (and maybe in `src/librustc/lib.rs`).

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks good! One further suggestion.

File: src/librustc_mir/transform/nll.rs (hunk 4 above)
Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: @nikomatsakis perfect, yeah I was already doing that as a suggestion of @arielb1. There are also a lot of Idx types that implement new and index methods twice, so I was removing copies of that.

Anyway thanks for the instructions here, it will probably help me a lot :).

Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: BTW: what do we do with `RustcEncodable` and `RustcDecodable` do we need to derive that?. If we really need I'd need to add an `extern crate` entry on the top level lib.rs because it's not included.

Comment (nikomatsakis) [reply src/librustc_mir/transform/nll.rs]: you can derive them -- you may need to add `extern crate serialize as rustc_serialize;` into the `lib.rs` if it's not already there

Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: Done in https://github.com/rust-lang/rust/pull/44878/commits/bea496666b518f26a8fef125fd907f51983e95dc and https://github.com/rust-lang/rust/pull/44878/commits/e266c0f7af963ea5177f7ce87f5e797140c69e26

BUT ... there are some things still pending and I have some questions:

1. I don't like that we need to add a lot of `use` clauses like in https://github.com/rust-lang/rust/pull/44878/commits/e266c0f7af963ea5177f7ce87f5e797140c69e26#diff-5d719cdea9e30d4d425f7b673017e897R22 Not sure what's the common practice in Rust, but wouldn't be better if we use the full path to lookup names inside the macro so we can avoid `use` outside?
2. I was starting to make all the code use the macro but in cases like https://github.com/rust-lang/rust/blob/master/src/librustc/dep_graph/graph.rs#L583 we need index to be public and I can't use ::new because it's a constant. Unsure what would be better to do there.
3. Not sure what to use as `debug_name` in cases like https://github.com/rust-lang/rust/pull/44878/commits/e266c0f7af963ea5177f7ce87f5e797140c69e26#diff-5d719cdea9e30d4d425f7b673017e897R160 and if I continue to migrate the rest of the code a lot of cases like that will shop up. I can just propose something and we can review later I guess.

Comment (nikomatsakis) [reply src/librustc_mir/transform/nll.rs]: > wouldn't be better if we use the full path to lookup names inside the macro

Yes, sorry, I missed that. Indeed we should be using the full path, in fact prefixed with `$crate` if we want to refer to something inside of `rustc_data_structures` (or `std` otherwise).

> we need index to be public \[to make constants\]

Hmm. So we could also use a `const fn` declared on the index type, but I think it would have to be an inherent method, since we don't support those yet in traits. 

> Not sure what to use as `debug_name`...

I think I'd prefer to remove that constant altogether and just use the name of the type. This can be extracted via "reflection" like so `unsafe { ::std::intrinsics::type_name::<T>() }`, which returns an `&'static str` with the name of the type `T`.

Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: I've fixed all this. I couldn't reuse `newtype_index` everywhere because of different issues.
I did a gist with my tries trying to convert some indexes to use the macro.

1. https://gist.github.com/anonymous/341c6951962f1ca62da32fb4c99d51c3
with this patch I'm getting ...
```
   Compiling rustc_plugin v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_plugin)
error: internal compiler error: src/librustc_const_eval/_match.rs:246: bad constructor ConstantValue(Struct({index(402): Integral(U32(0))})) for adt rustc::mir::Local

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.21.0-beta.1 (198109911 2017-08-29) running on x86_64-apple-darwin

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:490:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `rustc_mir`.

Caused by:
  process didn't exit successfully: `/Users/santiago/src/personal/rust/build/bootstrap/debug/rustc --crate-name rustc_mir src/librustc_mir/lib.rs --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=dac20bffaad97e2c -C extra-filename=-dac20bffaad97e2c --out-dir /Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps --target x86_64-apple-darwin -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/release/deps --extern bitflags=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/libbitflags-c9c6946b58ad8fa6.rlib --extern rustc_const_math=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/librustc_const_math-51531a5a4f8c3d3b.dylib --extern rustc_errors=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/librustc_errors-0699d9bc11181c8d.dylib --extern graphviz=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/libgraphviz-d8b388687c513e89.dylib --extern rustc_const_eval=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/librustc_const_eval-531305a0faa8fa56.dylib --extern rustc=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/librustc-936f28e604c3cd79.dylib --extern serialize=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/libserialize-4a223d9ec894bec3.dylib --extern serialize=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/libserialize-4a223d9ec894bec3.rlib --extern rustc_data_structures=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/librustc_data_structures-6bfac4716e737892.dylib --extern syntax_pos=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/libsyntax_pos-00853d8a91369014.dylib --extern syntax=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/libsyntax-b7ec86d4dc418fb0.dylib --extern log=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps/liblog-cc7339a977b3dc68.rlib -L native=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/build/miniz-sys-3abc8335811851be/out` (exit code: 101)
warning: build failed, waiting for other jobs to finish...
error: build failed
thread 'main' panicked at 'command did not execute successfully: "/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "build" "--target" "x86_64-apple-darwin" "-j" "4" "--release" "--features" " jemalloc llvm" "--manifest-path" "/Users/santiago/src/personal/rust/src/rustc/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101', src/bootstrap/compile.rs:883:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failed to run: /Users/santiago/src/personal/rust/build/bootstrap/debug/bootstrap build --incremental --stage 1 src/libstd
Build completed unsuccessfully in 0:04:03
```
unsure why.

2. https://gist.github.com/anonymous/369cc785baa07a54ca69e67b236878b2
in this case I need to make new pub, but if I do, I get errors for making pub something I don't need to be pub in other indexes.

Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: We would also need to check if we want this version https://github.com/rust-lang/rust/pull/44878/files#diff-8d8928737ae7efd735f00ba8e881c381L587, this version https://github.com/rust-lang/rust/pull/44878/files#diff-ad7502063d3dee7dc215765da3e83c58L25 or this https://github.com/rust-lang/rust/pull/44878/files#diff-fe3798f621a7c28fa0c90f86100fa421R58 version.

Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: And we also need to decide if we really want to change this https://github.com/rust-lang/rust/pull/44878/files#diff-f0577ac900ffbd36d3bb3421a928cbbdL565.
See also https://gitter.im/rust-impl-period/WG-compiler-nll?at=59ce446d210ac2692088280f

Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: After deciding and addressing this last comments, I think we are fine to make this thing mergeable again.

**Comment by nikomatsakis** [comment]
@spastorino @Nashenas88 needs rebase :(

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc_mir/transform/nll.rs (hunk 4 above)
Comment (nikomatsakis) [reply src/librustc_mir/transform/nll.rs]: OK so:

1. With respect to [this gist](https://gist.github.com/anonymous/341c6951962f1ca62da32fb4c99d51c3), I don't quite understand why that gist makes the macro take a configurable max. That said, I also don't know why it's causing an ICE. =) Would take some investigation.

2. For [this gist](https://gist.github.com/anonymous/369cc785baa07a54ca69e67b236878b2), what errors do you get? And what exactly did yo make `pub`?

3. Can you summarize the differences between those three gists?

4. I think we want the "short form"

**Review by nikomatsakis — APPROVED** [review]
Reviewing this, I think it's good enough to land.

File: src/librustc_mir/transform/nll.rs (hunk 4 above)
Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: 1. Because this case https://gist.github.com/anonymous/341c6951962f1ca62da32fb4c99d51c3#file-first_statement-patch-L95 needs a different max. The correct macro should take an optional u32, if provided use that if not use u32::MAX.
2. I'm getting ...
```
error[E0449]: unnecessary visibility qualifier
    --> src/librustc/mir/mod.rs:1512:1
     |
1512 | newtype_index!(Promoted, "promoted");
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `pub` not needed here
```
Because I'm making `pub` this https://gist.github.com/anonymous/369cc785baa07a54ca69e67b236878b2#file-crate_num-patch-L66 and https://gist.github.com/anonymous/369cc785baa07a54ca69e67b236878b2#file-crate_num-patch-L71
And if you don't make that `pub` you end getting a bunch of ...
```
error[E0599]: no function or associated item named `new` found for type `hir::def_id::CrateNum` in the current scope
   --> src/librustc/middle/dependency_format.rs:404:24
    |
404 |             let cnum = CrateNum::new(i + 1);
    |                        ^^^^^^^^^^^^^
    |
```
3. is what we talked yesterday, we went with `assert!(idx < ::std::u32::MAX as usize);`.
Now that I force pushed links does not make sense anymore
4. Yep, we ended discussing yesterday in gitter. I pushed code with the short version.

Comment (nikomatsakis) [reply src/librustc_mir/transform/nll.rs]: OK, some further thoughts:

1. Ok. I have no problem with a configurable max, just wanted to understand. I think perhaps we should also make the macro more descriptive, but that's a separate thing. e.g something like:

```rust
newtype_index!(TypeName, {
    max: MAXIMUM, // <-- these lines are optional
    description: "TypeName",
})
```

It'll take some macro fun to make it work =) but it'd be much prettier. 

2. Regarding the `pub` qualifier: indeed you cannot declare things to be public in trait impls. Those errors you are getting are likely because the `Idx` trait is not imported into scope at the points where you are writing `CrateNum::new`. This is an ergonomic annoyance of using the trait method `new` instead of the older approach based on inherent methods. One way to do both would be to have the macro generate an inherent method as well. So something like:

```rust
impl $name { // <-- NB: NOT `impl Idx for $name`
    pub fn new(value: usize) -> Self {
        assert!(value < (::std::u32::MAX) as usize);
        $name(value as u32)
    }

    pub fn index(self) -> usize {
        self.0 as usize
    }
}

impl Idx for $name {
    fn new(value: usize) -> Self { Self::new(value) }
    fn index(self) -> usize { self.index() }
}
```

The code at the very end might look surprising: what's happening there is that we are invoking the *inherent methods* from the *trait methods*. Though both have the same name, they are distinct to the compiler, and inherent methods get precedence.

Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: 1. sounds good. We would just need to figure out the ICE.
2. yep, I saw that this is the issue. Will check this and your suggestion is maybe better :).

Comment (spastorino) [reply src/librustc_mir/transform/nll.rs]: New patch https://gist.github.com/spastorino/4c9c55df5f28293bc131b9bfb576c41e that gives another ICE ...

```
~/src/personal/rust git:(nashenas88) ✗ ./x.py build --stage 1 src/libstd
Updating submodules
extracting /Users/santiago/src/personal/rust/build/cache/2017-08-29/rust-std-beta-x86_64-apple-darwin.tar.gz
extracting /Users/santiago/src/personal/rust/build/cache/2017-08-29/rustc-beta-x86_64-apple-darwin.tar.gz
extracting /Users/santiago/src/personal/rust/build/cache/2017-08-29/cargo-beta-x86_64-apple-darwin.tar.gz
    Finished dev [unoptimized] target(s) in 0.0 secs
Building stage0 std artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling core v0.0.0 (file:///Users/santiago/src/personal/rust/src/libcore)
   Compiling unwind v0.0.0 (file:///Users/santiago/src/personal/rust/src/libunwind)
   Compiling cc v1.0.0
   Compiling libc v0.2.31
   Compiling filetime v0.1.12
   Compiling build_helper v0.1.0 (file:///Users/santiago/src/personal/rust/src/build_helper)
   Compiling alloc_jemalloc v0.0.0 (file:///Users/santiago/src/personal/rust/src/liballoc_jemalloc)
   Compiling std v0.0.0 (file:///Users/santiago/src/personal/rust/src/libstd)
   Compiling compiler_builtins v0.0.0 (file:///Users/santiago/src/personal/rust/src/rustc/compiler_builtins_shim)
   Compiling cmake v0.1.26
   Compiling rustc_asan v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_asan)
   Compiling rustc_tsan v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_tsan)
   Compiling libc v0.0.0 (file:///Users/santiago/src/personal/rust/src/rustc/libc_shim)
   Compiling rand v0.0.0 (file:///Users/santiago/src/personal/rust/src/librand)
   Compiling std_unicode v0.0.0 (file:///Users/santiago/src/personal/rust/src/libstd_unicode)
   Compiling panic_abort v0.0.0 (file:///Users/santiago/src/personal/rust/src/libpanic_abort)
   Compiling alloc v0.0.0 (file:///Users/santiago/src/personal/rust/src/liballoc)
   Compiling collections v0.0.0 (file:///Users/santiago/src/personal/rust/src/libcollections)
   Compiling alloc_system v0.0.0 (file:///Users/santiago/src/personal/rust/src/liballoc_system)
   Compiling panic_unwind v0.0.0 (file:///Users/santiago/src/personal/rust/src/libpanic_unwind)
    Finished release [optimized] target(s) in 60.49 secs
Copying stage0 std from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Building stage0 test artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling getopts v0.2.15
   Compiling term v0.0.0 (file:///Users/santiago/src/personal/rust/src/libterm)
   Compiling test v0.0.0 (file:///Users/santiago/src/personal/rust/src/libtest)
    Finished release [optimized] target(s) in 12.53 secs
Copying stage0 test from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Building stage0 compiler artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling rustc_platform_intrinsics v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_platform_intrinsics)
   Compiling rustc-demangle v0.1.5
   Compiling cc v1.0.0
   Compiling graphviz v0.0.0 (file:///Users/santiago/src/personal/rust/src/libgraphviz)
   Compiling syntax v0.0.0 (file:///Users/santiago/src/personal/rust/src/libsyntax)
   Compiling rustc_back v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_back)
   Compiling rustc-serialize v0.3.24
   Compiling rustc_metadata v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_metadata)
   Compiling rustc_trans v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_trans)
   Compiling fmt_macros v0.0.0 (file:///Users/santiago/src/personal/rust/src/libfmt_macros)
   Compiling libc v0.2.31
   Compiling rustc_driver v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_driver)
   Compiling serialize v0.0.0 (file:///Users/santiago/src/personal/rust/src/libserialize)
   Compiling stable_deref_trait v1.0.0
   Compiling rustc_incremental v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_incremental)
   Compiling ar v0.3.0
   Compiling log v0.3.8
   Compiling rustc v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc)
   Compiling bitflags v1.0.0
   Compiling arena v0.0.0 (file:///Users/santiago/src/personal/rust/src/libarena)
   Compiling filetime v0.1.12
   Compiling miniz-sys v0.1.10
   Compiling owning_ref v0.3.3
   Compiling num_cpus v1.6.2
   Compiling jobserver v0.1.6
   Compiling env_logger v0.4.3
   Compiling rustc_cratesio_shim v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_cratesio_shim)
   Compiling build_helper v0.1.0 (file:///Users/santiago/src/personal/rust/src/build_helper)
   Compiling rls-span v0.4.0
   Compiling rustc_data_structures v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_data_structures)
   Compiling rustc_apfloat v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_apfloat)
   Compiling rls-data v0.10.0
   Compiling rustc_llvm v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_llvm)
   Compiling syntax_pos v0.0.0 (file:///Users/santiago/src/personal/rust/src/libsyntax_pos)
   Compiling flate2 v0.2.20
   Compiling rustc_errors v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_errors)
   Compiling rustc_const_math v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_const_math)
   Compiling proc_macro v0.0.0 (file:///Users/santiago/src/personal/rust/src/libproc_macro)
   Compiling syntax_ext v0.0.0 (file:///Users/santiago/src/personal/rust/src/libsyntax_ext)
   Compiling rustc_resolve v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_resolve)
   Compiling rustc_allocator v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_allocator)
   Compiling rustc_trans_utils v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_trans_utils)
   Compiling rustc_const_eval v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_const_eval)
   Compiling rustc_typeck v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_typeck)
   Compiling rustc_lint v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_lint)
   Compiling rustc_mir v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_mir)
   Compiling rustc_passes v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_passes)
   Compiling rustc_save_analysis v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_save_analysis)
   Compiling rustc_privacy v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_privacy)
   Compiling rustc_borrowck v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_borrowck)
   Compiling rustc_plugin v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_plugin)
   Compiling rustc-main v0.0.0 (file:///Users/santiago/src/personal/rust/src/rustc)
    Finished release [optimized] target(s) in 931.86 secs
Copying stage0 rustc from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Assembling stage1 compiler (x86_64-apple-darwin)
Building stage1 std artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling libc v0.2.31
   Compiling core v0.0.0 (file:///Users/santiago/src/personal/rust/src/libcore)
   Compiling cc v1.0.0
   Compiling unwind v0.0.0 (file:///Users/santiago/src/personal/rust/src/libunwind)
   Compiling filetime v0.1.12
   Compiling build_helper v0.1.0 (file:///Users/santiago/src/personal/rust/src/build_helper)
   Compiling std v0.0.0 (file:///Users/santiago/src/personal/rust/src/libstd)
   Compiling cmake v0.1.26
   Compiling compiler_builtins v0.0.0 (file:///Users/santiago/src/personal/rust/src/rustc/compiler_builtins_shim)
   Compiling alloc_jemalloc v0.0.0 (file:///Users/santiago/src/personal/rust/src/liballoc_jemalloc)
   Compiling rustc_tsan v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_tsan)
   Compiling rustc_asan v0.0.0 (file:///Users/santiago/src/personal/rust/src/librustc_asan)
   Compiling std_unicode v0.0.0 (file:///Users/santiago/src/personal/rust/src/libstd_unicode)
   Compiling libc v0.0.0 (file:///Users/santiago/src/personal/rust/src/rustc/libc_shim)
   Compiling rand v0.0.0 (file:///Users/santiago/src/personal/rust/src/librand)
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.22.0-dev running on x86_64-apple-darwin

thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:335:20
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `rand`.

Caused by:
  process didn't exit successfully: `/Users/santiago/src/personal/rust/build/bootstrap/debug/rustc --crate-name rand src/librand/lib.rs --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=81b41dd92bfaa446 -C extra-filename=-81b41dd92bfaa446 --out-dir /Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps --target x86_64-apple-darwin -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/release/deps --extern core=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libcore-8564d565e8172c07.rlib` (exit code: 101)
warning: build failed, waiting for other jobs to finish...
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.22.0-dev running on x86_64-apple-darwin

thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:335:20
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `compiler_builtins`.

Caused by:
  process didn't exit successfully: `/Users/santiago/src/personal/rust/build/bootstrap/debug/rustc --crate-name compiler_builtins src/rustc/compiler_builtins_shim/../../libcompiler_builtins/src/lib.rs --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 --cfg feature="c" --cfg feature="compiler-builtins" --cfg feature="default" --cfg feature="rustbuild" -C metadata=abe9981e0d697dbc -C extra-filename=-abe9981e0d697dbc --out-dir /Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps --target x86_64-apple-darwin -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/release/deps --extern core=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libcore-8564d565e8172c07.rlib -L native=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/build/compiler_builtins-0375618ff04bf6f3/out -l static=compiler-rt` (exit code: 101)
warning: build failed, waiting for other jobs to finish...
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.22.0-dev running on x86_64-apple-darwin

thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:335:20
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.22.0-dev running on x86_64-apple-darwin

thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:335:20
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `std_unicode`.

Caused by:
  process didn't exit successfully: `/Users/santiago/src/personal/rust/build/bootstrap/debug/rustc --crate-name std_unicode src/libstd_unicode/lib.rs --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=07935fec0d990c43 -C extra-filename=-07935fec0d990c43 --out-dir /Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps --target x86_64-apple-darwin -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/release/deps --extern core=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libcore-8564d565e8172c07.rlib` (exit code: 101)
warning: build failed, waiting for other jobs to finish...
error: Could not compile `libc`.

Caused by:
  process didn't exit successfully: `/Users/santiago/src/personal/rust/build/bootstrap/debug/rustc --crate-name libc src/rustc/libc_shim/../../liblibc/src/lib.rs --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 --cfg feature="default" --cfg feature="stdbuild" -C metadata=193d229c9d522195 -C extra-filename=-193d229c9d522195 --out-dir /Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps --target x86_64-apple-darwin -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps -L dependency=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/release/deps --extern core=/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libcore-8564d565e8172c07.rlib` (exit code: 101)
thread 'main' panicked at 'command did not execute successfully: "/Users/santiago/src/personal/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "build" "--target" "x86_64-apple-darwin" "-j" "4" "--release" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/Users/santiago/src/personal/rust/src/libstd/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101', src/bootstrap/compile.rs:883:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failed to run: /Users/santiago/src/personal/rust/build/bootstrap/debug/bootstrap build --stage 1 src/libstd
Build completed unsuccessfully in 0:17:31
```
